From 1b74b4ffff02c1f824b5b26ed0576770d07a9605 Mon Sep 17 00:00:00 2001
From: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Date: Mon, 28 Nov 2022 01:22:18 -0800
Subject: [PATCH 2/2] ath12k: Avoid register address caculations on a per
 packet basis

Precompute the datapath registers addresses and use it directly to save
cpu time. Do not compute the addresses on a per packet basis.

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c |  4 ++--
 drivers/net/wireless/ath/ath12k/hal.c   | 13 +++++++++++++
 drivers/net/wireless/ath/ath12k/hal.h   | 11 ++++-------
 drivers/net/wireless/ath/ath12k/mac.c   |  5 ++---
 4 files changed, 21 insertions(+), 12 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -81,7 +81,7 @@ struct ath12k_tx_desc_info *ath12k_dp_tx
 					struct ath12k_tx_desc_info,
 					list);
 	if (!desc) {
-		ath12k_warn(dp->ab, "failed to allocate data Tx buffer\n");
+		ath12k_dbg(dp->ab, ATH12K_DBG_DP_TX, "failed to allocate data Tx desc\n");
 		return NULL;
 	}
 
@@ -120,7 +120,6 @@ int ath12k_dp_tx_direct(struct ath12k_vi
 	struct dp_tx_ring *tx_ring;
 	struct device *dev = arvif->dev;
 	void *hal_tcl_desc;
-	void *mem = ab->mem;
 	struct hal_tcl_data_cmd tcl_desc;
 	u8 hal_ring_id, ring_id;
 	dma_addr_t paddr;
@@ -159,14 +158,14 @@ int ath12k_dp_tx_direct(struct ath12k_vi
 	hal_ring_id = ring_id + HAL_SRNG_RING_ID_SW2TCL1;
 	tcl_ring = &ab->hal.srng_list[hal_ring_id];
 
-	ath12k_hal_srng_access_src_ring_begin_nolock(ab, tcl_ring);
+	ath12k_hal_srng_access_src_ring_begin_nolock(tcl_ring);
 
 	hal_tcl_desc = (void *)ath12k_hal_srng_src_get_next_entry_nolock(ab, tcl_ring);
 	if (unlikely(!hal_tcl_desc)) {
 		/* NOTE: It is highly unlikely we'll be running out of tcl_ring
 		 * desc because the desc is directly enqueued onto hw queue.
 		 */
-		ath12k_hal_srng_access_umac_src_ring_end_nolock(mem, tcl_ring);
+		ath12k_hal_srng_access_umac_src_ring_end_nolock(tcl_ring);
 		ab->soc_stats.tx_err.desc_na[ring_id]++;
 		ret = -ENOMEM;
 		goto fail_unmap_dma;
@@ -185,7 +184,7 @@ int ath12k_dp_tx_direct(struct ath12k_vi
 	tcl_desc.info5 = 0;
 	memcpy(hal_tcl_desc, &tcl_desc, sizeof(tcl_desc));
 	dsb(st);
-	ath12k_hal_srng_access_umac_src_ring_end_nolock(mem, tcl_ring);
+	ath12k_hal_srng_access_umac_src_ring_end_nolock(tcl_ring);
 
 	atomic_inc(&ar->dp.num_tx_pending);
 
@@ -238,7 +237,7 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 	tx_desc = ath12k_dp_tx_assign_buffer_nolock(dp, ti.ring_id);
 	if (unlikely(!tx_desc)) {
 		ab->soc_stats.tx_err.txbuf_na[ti.ring_id]++;
-		return -ENOSPC;
+		return -ENOMEM;
 	}
 
 	ti.bank_id = arvif->bank_id;
@@ -363,13 +362,13 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 	hal_ring_id = tx_ring->tcl_data_ring.ring_id;
 	tcl_ring = &ab->hal.srng_list[hal_ring_id];
 
-	ath12k_hal_srng_access_src_ring_begin_nolock(ab, tcl_ring);
+	ath12k_hal_srng_access_src_ring_begin_nolock(tcl_ring);
 	hal_tcl_desc = (void *)ath12k_hal_srng_src_get_next_entry(ab, tcl_ring);
 	if (unlikely(!hal_tcl_desc)) {
 		/* NOTE: It is highly unlikely we'll be running out of tcl_ring
 		 * desc because the desc is directly enqueued onto hw queue.
 		 */
-		ath12k_hal_srng_access_umac_src_ring_end_nolock(ab->mem, tcl_ring);
+		ath12k_hal_srng_access_umac_src_ring_end_nolock(tcl_ring);
 		ab->soc_stats.tx_err.desc_na[ti.ring_id]++;
 		ret = -ENOMEM;
 
@@ -388,7 +387,7 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 	tcl_cmd->info4 = arvif->desc.info4;
 	tcl_cmd->info5 = 0;
 	dsb(st);
-	ath12k_hal_srng_access_umac_src_ring_end_nolock(ab->mem, tcl_ring);
+	ath12k_hal_srng_access_umac_src_ring_end_nolock(tcl_ring);
 
 	ath12k_dbg_dump(ab, ATH12K_DBG_DP_TX, NULL, "dp tx msdu: ",
 			skb->data, skb->len);
@@ -857,7 +856,7 @@ int ath12k_dp_tx_completion_handler(stru
 
 	valid_entries = ath12k_hal_srng_dst_num_free(ab, status_ring, false);
 	if (!valid_entries) {
-		ath12k_hal_srng_access_umac_dst_ring_end_nolock(ab->mem, status_ring);
+		ath12k_hal_srng_access_umac_dst_ring_end_nolock(status_ring);
 		return 0;
 	}
 
@@ -951,7 +950,7 @@ int ath12k_dp_tx_completion_handler(stru
 		}
 	}
 
-	ath12k_hal_srng_access_umac_dst_ring_end_nolock(ab->mem, status_ring);
+	ath12k_hal_srng_access_umac_dst_ring_end_nolock(status_ring);
 	dev_kfree_skb_list_fast(&free_list_head);
 
 	return (orig_budget - budget);
--- a/drivers/net/wireless/ath/ath12k/hal.c
+++ b/drivers/net/wireless/ath/ath12k/hal.c
@@ -1371,6 +1371,12 @@ int ath12k_hal_srng_setup(struct ath12k_
 		if (srng_config->mac_type == ATH12K_HAL_SRNG_UMAC) {
 			srng->u.src_ring.hp_addr =
 				(u32 *)((unsigned long)ab->mem + reg_base);
+			if (type  == HAL_TCL_DATA) {
+				srng->u.src_ring.hp_addr_direct =
+					(u32 *)((unsigned long)ab->mem +
+					HAL_DP_REG_WINDOW_OFFSET +
+					(reg_base & WINDOW_RANGE_MASK));
+			}
 		} else {
 			idx = ring_id - HAL_SRNG_RING_ID_DMAC_CMN_ID_START;
 			srng->u.src_ring.hp_addr = (void *)(hal->wrp.vaddr +
@@ -1394,6 +1400,13 @@ int ath12k_hal_srng_setup(struct ath12k_
 			srng->u.dst_ring.tp_addr =
 				(u32 *)((unsigned long)ab->mem + reg_base +
 				(HAL_REO1_RING_TP - HAL_REO1_RING_HP));
+			if (type  == HAL_WBM2SW_RELEASE) {
+				srng->u.dst_ring.tp_addr_direct =
+					(u32 *)((unsigned long)ab->mem +
+					(reg_base & WINDOW_RANGE_MASK) +
+					HAL_DP_REG_WINDOW_OFFSET +
+					(HAL_REO1_RING_TP - HAL_REO1_RING_HP));
+			}
 		} else {
 			/* For PMAC & DMAC rings, tail pointer updates will be done
 			 * through FW by writing to a shared memory location
--- a/drivers/net/wireless/ath/ath12k/hal.h
+++ b/drivers/net/wireless/ath/ath12k/hal.h
@@ -728,6 +728,7 @@ struct hal_srng {
 			 * accessed through SW structure
 			 */
 			u32 *tp_addr;
+			u32 *tp_addr_direct;
 
 			/* Current SW loop cnt */
 			u32 loop_cnt;
@@ -757,6 +758,7 @@ struct hal_srng {
 			 * through SW structure
 			 */
 			u32 *hp_addr;
+			u32 *hp_addr_direct;
 
 			/* Low threshold - in number of ring entries */
 			u32 low_threshold;
@@ -1207,8 +1209,7 @@ u32 *ath12k_hal_srng_src_get_next_entry_
 }
 
 static inline
-void ath12k_hal_srng_access_src_ring_begin_nolock(void *mem,
-						  struct hal_srng *srng)
+void ath12k_hal_srng_access_src_ring_begin_nolock(struct hal_srng *srng)
 {
 	srng->u.src_ring.cached_tp = *(volatile u32 *)srng->u.src_ring.tp_addr;
 }
@@ -1224,26 +1225,19 @@ void ath12k_hal_srng_access_dst_ring_beg
 }
 
 static inline
-void ath12k_hal_srng_access_umac_src_ring_end_nolock(void *mem,
-						     struct hal_srng *srng)
+void ath12k_hal_srng_access_umac_src_ring_end_nolock(struct hal_srng *srng)
 {
-	u32 offset = (unsigned long)srng->u.src_ring.hp_addr - (unsigned long)mem;
-
 	srng->u.src_ring.last_tp =
 			*(volatile u32 *)srng->u.src_ring.tp_addr;
-	writel_relaxed(srng->u.src_ring.hp, mem + HAL_DP_REG_WINDOW_OFFSET
-		  + (offset & WINDOW_RANGE_MASK));
+	writel_relaxed(srng->u.src_ring.hp, srng->u.src_ring.hp_addr_direct);
 	srng->timestamp = jiffies;
 }
 
 static inline
-void ath12k_hal_srng_access_umac_dst_ring_end_nolock(void *mem,
-						     struct hal_srng *srng)
+void ath12k_hal_srng_access_umac_dst_ring_end_nolock(struct hal_srng *srng)
 {
-	u32 offset = (unsigned long)srng->u.dst_ring.tp_addr - (unsigned long)mem;
 	srng->u.dst_ring.last_hp = *srng->u.dst_ring.hp_addr;
-	writel_relaxed(srng->u.dst_ring.tp, mem + HAL_DP_REG_WINDOW_OFFSET
-		       + (offset & WINDOW_RANGE_MASK));
+	writel_relaxed(srng->u.dst_ring.tp, srng->u.dst_ring.tp_addr_direct);
 	srng->timestamp = jiffies;
 }
 
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -6867,9 +6867,8 @@ static void ath12k_mac_op_tx(struct ieee
 	if (skb->fast_xmit) {
 		ret = ath12k_dp_tx_direct(arvif, skb);
 		if (unlikely(ret)) {
-			ar = arvif->ar;
-			ath12k_warn(ar->ab, "failed to transmit frame %d\n",
-				    ret);
+			ath12k_dbg(arvif->ar->ab, ATH12K_DBG_MAC,
+				   "failed to transmit frame %d\n", ret);
 			dev_kfree_skb_any(skb);
 		}
 		return;
